### ISSUE TITLE
Allow non-nullable optional arg for generic type

### ISSUE THREAD

**Issue by lesnitsky** [dart-lang/sdk#51708]
Functions like `void foo<T>([T optionalArg])` could only have `optionalArg` as nullable (`T?`). 

Optional arg could only be non-nullable if there is a known const default value, but this is never the case for generics.

I've tried something like this as a workaround:

```dart
void foo<T extends Object>([T optionalArg.= const Object() as T]) {}
```

The reasoning behind that workaround is "whenever the arg is not provided – I'm ok to downcast `T` to `Object`.

Analyzer didn't give any warnings about this code, however, build failed with the following error:

```
void foo<T extends Object>([T value = const Object() as T]) {}
                                                     ^
Target kernel_snapshot failed: Exception
```

This could have been solved with a default types for type arguments, so something like this:

```
void foo<T extends Object = Object>([T dep = const Object]) {}
```
Are there any plans on adding default types? What would be a language construct that indicates that something is optional to provide, but when provided – should not be null?

**Comment by lesnitsky** [comment]
I believe the same is relevant to non-required named args

**Comment by lrhn** [comment]
Having a default type argument doesn't help, a `const Object()` still cannot be a default value if `T` is actually `int`.
As you say, there is no one value which can satisfy all type arguments passed as `T`, because `Never` is a valid type argument, and it has no values at all.

There is currently no other alternative than using `T?` as parameter type.
Also, if `<T extends Object>`, that's a perfectly reasonable implementation, which makes it clear that `null` represents no value.

The one alternative is to make parameters not be optional if the default default value doesn't match the parameter's type.
(I think I have a proposal for it, but I can't find it right now).

**Comment by lesnitsky** [comment]
> make parameters not be optional if the default default value doesn't match the parameter's type.

I'm not sure I understand this bit, maybe if you could find the proposal you mentioned it will become more clear

**Comment by lesnitsky** [comment]
> Having a default type argument doesn't help, a const Object() still cannot be a default value if T is actually int.

This is true when `T` is explicitly defined, so `foo` is being used as `foo<int>()`

My problem (and a workaround that didn't work) is for cases, where `T` is being inferred from what was passed as an argument, so if nothing is passed – `T == Object`, so `const Object()` is a correct default value, but for this "fallback" to work, there should be some compiler/analyzer level functionality that forces an arg to be provided if `T` is explicitly defined.

I believe there are plenty of cases where named not-required args could also have a default value of a lower bound of the generic type.

**Comment by lrhn** [comment]
The problem here is that the type of your function is `void Function<T extends Object>([T])`. At the point where it's called, that's all the compiler knows. It does not know about the default value, so all it can see is that you can call `foo` with no argument, and it will infer a `T` value from that, which can be anything, because `T` is unconstrained. It might just be `Object`, and that would work.
*But* you can also call it as `foo<int>()`, and according to the type, that's a valid invocation. And it cannot work.

At the declaration point, `void foo<T extends Object>([T o = const Object()])` cannot be allowed because it would be unsound to allow that function to be called as `foo<int>()`, but there is nothing in the type system which prevents that.
And the default value must be *valid* for that invocation, which it isn't.

So what we really need is:
* Allow default values which are not valid for all type parameters.

That requires us to figure out what happens when it's not valid. The obvious behavior is that then the parameter is not optional any more. Which means that we have parameters which are *conditionally optional*, based on whether the default value type satisfies the parameter type.
I'm sure I've specified that before somewhere, but I can't find it now, so here we go:

## Conditionally optional parameters

A conditionally optional parameter is an optional parameter with a default value (type) which is not guaranteed to satisfy the parameter type. _Which mans that the parameter type must contain at least one type parameter, otherwise it's statically decidable whether it satisfies the type or not._

To allow this, we introduce *default value types* on type parameters in function types, to carry the information from declaration point to call point:
```dart
(T, T) createPoint<T>({T x = 0, T y = 0}) => (x, y);
```
will have the function type:
```dart
  (T, T) Function<T>({T x = int, T y = int})
```
The `= int` is the type of the default value.
That means that if `T` is `int` or `num` (or `Object` or any top type, or `FutureOr` of any of those), the arguments to `x` and `y` are optional. If not, you have to supply a value which satisfies the actual `T`.

If a function has a default value which is always a subtype of the parameter type, it is not part of the function type.
That is, we cannot distinguish `void Function([num x = int])` and `void Function([num x = double])` as function types, they are both just `void Function([num x])`. Safe for `void foo<T>([List<T> l = const <Never>[]])`, where we won't remember the  default value type, it's just `void Function<T>([List<T>])`.

We define subtyping on such function types ... somehow. (Likely: parameter of `T1 = D1` can substitute for a parameter `T2 = D2` if `T2` \<: `T1` and `D1` \<: `D2`. A conditionally optional parameter can always substitute for a required parameter. An optional parameter with no default value type can count as having a default value type of `Never`, which satisfies every parameter type.)

A function having no default value means having the default value `null`. An optional parameter with no default value used to have to be be nullable, not it can be conditionally optional instead, with a default value type of `Null` if the parameter itself isn't always nullable.
(That's not the same as a function *type* parameter having no default value *type*, which just means that it has a type which cannot fail.)

Override and implementation rules work as normal: An overriding subtype method must have a function signature which is a subtype of the supertype's method signature. The function type of the concrete method must satisfy the function signature of the interface it implements. 

We'll can make it a compile-time error if the default value *cannot* be a subtype of the parameter type, or just a warning,
or make it just count as equivalent to a required parameter.
Say `Function<T extends Object>([T = Null])`. That's just a long way to write a required parameter.

_And_ we can then also use the default value type as the type of the missing argument in inference, because we have the type at the invocation point.

Going back to the original problem, that makes your code be:
```dart
void foo<T extends Object>([T optionalArg = const Object()]) {}
```
which has type `void Function<T extends Object>([T = Object])`.

Invoking that with no type argument and no argument will use the default value, infer `T` to be `Object` from that, and just work.
If you do `foo(1)` it can infer `T` to be `int`. If you do `foo<int>()`, you get an error because the default value's type, `Object`, is not assignable to the actual parameter type, making the parameter effectively required.


## Consequences
This complicates function types, which means it complicates the type system. That's always a cost.
There will be people confused why they can't do 
```dart
T defaultValue<T>([T value = null]) => value;
int Function([int]) f = defaultValue<int>;
```
They cannot, because the function type `int Function([int = null])` cannot be used as `int Function([int])`, it's not substitutable.
The only assignment you'd get away with is `int Function(int) f = defaultValue<int>;`.

But if we make unsatisfiable parameters count as required, we may now have required parameters in the middle of a parameter list:
```dart
void foo<T>([int x = 0, T y = Object(), int z = 0]) { ... }
```
has an optional `x` and `z` parameter, but a conditionally optional `y` parameter.
(We can't solve this by requireing conditionally optional parameters to go before unconditionally optional parameters, because two separate conditionally optional parameters can still end up with the former optional and the latter not.)

So we need to handle optional parameters before required parameters. (I'd go with an elision like `foo(, 42)` to omit the `x` argument.)

So, there are complications. But if successful, it would solve some of the API problems we have in the post-null safety world,
like:
```dart
  factory Future.value([T value = null]) { ... }
```
would now only allow you to omit the argument *iff* `T` allows `null` as a value, otherwise it'd be required.
(Currently it's typed as `([T? value = null])` and throws at runtime if you omit the value and `T` isn't nullable.
That's the case that caused me to originally write out this idea.)
